sdks/python/apache_beam/runners/direct/executor.py

```diff
@@ -290,70 +293,87 @@ def __init__(self, transform_evaluator_registry, evaluation_context,
+        self._applied_ptransform.full_label,
+        'start',
+        metrics_container=metrics_container)
+    process_state = state_sampler.scoped_state(
+        self._applied_ptransform.full_label,
+        'process',
+        metrics_container=metrics_container)
+    finish_state = state_sampler.scoped_state(
+        self._applied_ptransform.full_label,
+        'finish',
+        metrics_container=metrics_container)
+
+    with start_state:
+      for side_input in self._applied_ptransform.side_inputs:
+        # Find the projection of main's window onto the side input's window.
+        window_mapping_fn = side_input._view_options().get(
+            'window_mapping_fn', sideinputs._global_window_mapping_fn)
+        main_onto_side_window = window_mapping_fn(
+            self._latest_main_input_window)
+        block_until = main_onto_side_window.end
+
+        if side_input not in self._side_input_values:
+          value = self._evaluation_context.get_value_or_block_until_ready(
+              side_input, self, block_until)
+          if not value:
+            # Monitor task will reschedule this executor once the side input is
+            # available.
+            return
+          self._side_input_values[side_input] = value
+      side_input_values = [
+          self._side_input_values[side_input]
+          for side_input in self._applied_ptransform.side_inputs]
+
+      while self._retry_count < self._max_retries_per_bundle:
+        try:
+          self.attempt_call(metrics_container,
+                            side_input_values,
+                            process_state,
+                            finish_state)
+          break
+        except Exception as e:
+          self._retry_count += 1
+          logging.error(
+              'Exception at bundle %r, due to an exception.\n %s',
+              self._input_bundle, traceback.format_exc())
+          if self._retry_count == self._max_retries_per_bundle:
+            logging.error('Giving up after %s attempts.',
+                          self._max_retries_per_bundle)
+            self._completion_callback.handle_exception(self, e)
 
     self._evaluation_context.metrics().commit_physical(
         self._input_bundle,
         metrics_container.get_cumulative())
     self._transform_evaluation_state.complete(self)
 
   def attempt_call(self, metrics_container,
-                   scoped_metrics_container,
-                   side_input_values):
+                   side_input_values,
+                   process_state,
+                   finish_state):
+    """Attempts to run a bundle. Called within the 'start' ExecutionState."""
     evaluator = self._transform_evaluator_registry.get_evaluator(
         self._applied_ptransform, self._input_bundle,
-        side_input_values, scoped_metrics_container)
+        side_input_values)
 
-    with scoped_metrics_container:
-      evaluator.start_bundle()
+    evaluator.start_bundle()
```

Comment:
Why is this not called from within start_state?

Reply:
The function `attempt_call` is already called within start_state (because side input initialization happens before `attempt_call` is called. I've documented this in the `attempt_call` docstring, but we could do it differently if you think that'd make more sense.

Reply:
Yes, I think it'd be cleaner to have all states explicitly entered within attempt_call than make this an implicit assumption about the caller.

Reply:
Done.